lib/generateClass.js

```diff
@@ -284,14 +307,222 @@ function writeToJson(def, cw) {
+ * @private
+ */
+function writeFromFhir(def, specs, fhir, cw) {
+
+  const fhirProfile = [...fhir.profiles, ...fhir._noDiffProfiles].find(p => p.id === fhirID(def.identifier));
+  const fhirExtension = fhir.extensions.find(p => p.id === fhirID(def.identifier, 'extension'));
+  const alreadyMappedElements = new Map;
+
+  cw.ln(`const inst = new ${className(def.identifier.name)}();`);
+
+  if(fhirProfile){
+    // FHIR profile exists, so we will pull out the mapping information from element maps
+
+    // first we do some pre-processing to identify slices. 
+    // slicing requires information spread across multiple elements so it doesn't play nice with the iterative approach below
+    // example: https://www.hl7.org/fhir/profiling-examples.html#blood-pressure
+    // say we want to pull out the "systolic" component, 
+    //  1) we need the context of the prior Observation.component element to know what the discriminator of the slice is ("code")
+    //  2) we need the fixed value of the later Obervation.component.code element to know which value to look for
+    // so to handle this we create a map of sliceGroup -> sliceName -> discriminator&code. in the blood pressure example, we'd produce a map like the following:
+    // { 'Observation.component' : {
+    //     'systolic': [{path: 'Observation.component.code', value: <loinc, 8480-6...>}], 
+    //     'diastolic': [{path: 'Observation.component.code', value: <loinc, 8462-4...>}] 
+    //   }
+    // }
+    // 
+    // to build up this map, we keep a stack because slices can be nested
+
+    const sliceMap = {};
+    const sliceStack = []; // note: javascript doesn't have Array.peek so we use shift/unshift to keep the "top" at index 0
+
+    for (let element of fhirProfile.snapshot.element) {
+
+      while (sliceStack.length > 0 && !element.path.startsWith(sliceStack[0].path)) {
+        // ended a slice group so pop the top element
+        const sliceGroup = sliceStack.shift();
+
+        sliceMap[sliceGroup.path] = sliceGroup.elements;
+      }
+
+      if (element.slicing && !element.path.endsWith('.extension')) { // extensions automatically slice on url, we don't care about that here
+        // starting a new slice group
+
+        const elementSlicing = JSON.parse(JSON.stringify(element.slicing)); // deep copy so we don't stomp on the source object
+
+        // pre-process and combine the element paths so we don't have to do it on every check
+        elementSlicing.discriminator.forEach( d => d.path = element.path + '.' + d.path);
+
+        const sliceGroup = { path: element.path, slicing: elementSlicing, currentSlice: null, elements: {} };
+
+        sliceStack.unshift(sliceGroup);
+      } else if (sliceStack[0]) {
+
+        if (element.sliceName) {
+          // starting a new slice element, only possible if already within a slice group
+
+          sliceStack[0].currentSlice = element.sliceName;
+          sliceStack[0].elements[element.sliceName] = [];
+
+        } else {
```

Comment:
Pretty much all of your comments here have been automatically marked as "outdated" because of significant code refactoring, other than this one. This one refers to the section which is now in function `preprocessSlicing` (around line 826 currently). I did move that section out of the else block, so I'm going to mark this as resolved, and try and find and double check the other comments.